fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1354,13 +1370,12 @@ def _post_reduction_hook(self, param: Parameter, reduced_grad: torch.Tensor) ->
     def _queue_wait_for_post_backward(self) -> None:
         """Try to queue a `wait_for_post_backward` callback.
 
-        Only called on root and only queue one callback. But can be called by
-        children FSDPs via a closure in case the root instance doesn't own any
-        params.
+        Only called on root and only queue one callback at the beginning of
+        outer most backward.
         """
         assert self._is_root
-        self.assert_state([TrainingState.BACKWARD_PRE, TrainingState.BACKWARD_POST])
-        if not self._post_backward_callback_queued:
+        if not self._post_backward_callback_queued and self._checkpoint_module_last_backward_call():
```

Comment:
maybe we can simplify things a bit by asserting root module is not checkpointed? It doesn't seem to make a lot of sense to checkpoint the root instance, it would not save any GPU memory. So we can just assert that during __init__, the root's module doesn't have a checkpoint counter attached to it and simplify the logic here?

Reply:
sounds good, let us explicitly not allow to checkpoint a root module!